src/core/diffusion/euler_grid.cc

```diff
@@ -143,4 +154,140 @@ void EulerGrid::DiffuseWithOpenEdge(real_t dt) {
   c1_.swap(c2_);
 }
 
+void EulerGrid::DiffuseWithDirichlet(real_t dt) {
+  const auto nx = resolution_;
+  const auto ny = resolution_;
+  const auto nz = resolution_;
+
+  const real_t ibl2 = 1 / (box_length_ * box_length_);
+  const real_t d = 1 - dc_[0];
+
+  const auto sim_time = GetSimulatedTime();
+
+  constexpr size_t YBF = 16;
+#pragma omp parallel for collapse(2)
+  for (size_t yy = 0; yy < ny; yy += YBF) {
+    for (size_t z = 0; z < nz; z++) {
+      size_t ymax = yy + YBF;
+      if (ymax >= ny) {
+        ymax = ny;
+      }
+      for (size_t y = yy; y < ymax; y++) {
+        size_t x{0};
+        size_t c{0};
+        size_t n{0};
+        size_t s{0};
+        size_t b{0};
+        size_t t{0};
+        c = x + y * nx + z * nx * ny;
+#pragma omp simd
+        for (x = 0; x < nx; x++) {
+          if (x == 0 || x == (nx - 1) || y == 0 || y == (ny - 1) || z == 0 ||
+              z == (nz - 1)) {
+            c2_[c] = boundary_condition_->Evaluate(x, y, z, sim_time);
+            ++c;
+            continue;
```

Comment:
I think I'd prefer
```
if ("box == boundary"){
  // treat boundary
} else {
  // Standard heat equation stencil
}
c++;
```
which would quite clearly differentiate between boundary and regular updates